### DISCUSSION TITLE
Where to place my stream queries

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#2365] [Q&A] [answered]
In my app, I'm using different queries in `StreamBuilder`s to get the data for the UI from the database. I'm reworking things a bit because the performance isn't great currently (due to many factors I think). My question is, what should be the default approach:

Say I have the following tables:
```sql
-- There are a number of cars
CREATE TABLE cars (
  id INT NOT NULL PRIMARY KEY,
  color TEXT
);

-- Each car has N wheels
CREATE TABLE wheels (
  id INT NOT NULL PRIMARY KEY,  
  car INT NULL REFERENCES cars(id),
  size REAL
);

-- Each wheel has N tires (maybe some summer and winter tires)
CREATE TABLE tires (
  id INT NOT NULL PRIMARY KEY,
  wheel INT NULL REFERENCES wheels(id),
  pressure REAL
)
```

Now in my app, I have a `Widget` to display a car, say the structure looks like this:
```yaml
- CarWidget
  - WheelWidget
    - TireWidget
    - TireWidget
  - WheelWidget
    - TireWidget
    - TireWidget
 - WheelWidget
    - TireWidget
    - TireWidget
 - WheelWidget
    - TireWidget
    - TireWidget
```
Would it now be recommended to have a `CarWidget` that listens to `db.select(db.cars)` and pass the `id` of each car down to the`WheelsWidget`, which uses it to select the wheels, and then in the `WheelsWidget` to pass down the `id` of each wheel to a `TiresWidget`, so that I end up with a single stream builder in each widget?

This is nice, because if a tire updates, not the whole widget will be rebuilt. Instead, all 8 tire widgets are built anew.

On the other hand, I could join the wheels and tires to the car table and only have the streambuilder in the CarWidget, just passing down data to the underlying Widgets.

What is the recommended way to approach such a case? Of course in reality, my model is a bit more complex, and there are cases where some queried tables are reused in some of the children but not in others, which may change the tradeoff made too.

**Comment by FaFre** [comment, marked as answer]
I think you want to make use of `watch()` that listens for changes and returns a handy stream.

**Reply by frankvollebregt** [reply]
Hi! Thanks for your response. I'm already using `watch()` to stream the queries, but the question is what the best approach is: should I always push all the stream builders down to a smaller level, but watch the same table multiple times in the process:
```yaml
CarWidget(
  watch(cars)
  WheelWidget(
    watch(wheels)
    TireWidget(
      watch(tires)
    )
  ),
  WheelWidget(
    watch(wheels)
    TireWidget(
      watch(tires)
    )
  ),
  WheelWidget(
    watch(wheels)
    TireWidget(
      watch(tires)
    )
  ),
  WheelWidget(
    watch(wheels)
    TireWidget(
      watch(tires)
    )
  )
)
```
Or should I only watch each table as few times as possible, and pass data down from the top widget:
```yaml
CarWidget(
  watch(carsWithWheelsWithTires)
  WheelWidget(
    TireWidget()
  )
  WheelWidget(
    TireWidget()
  )
  WheelWidget(
    TireWidget()
  )
  WheelWidget(
    TireWidget()
  )
)
```

**Reply by FaFre** [reply]
Since you listen everywhere to the same stream it would be best to rebuild everything and just have one subscription. Since a stream issues async events you would have elsewise rebuilds of the widgets individually , instead of only one, which will be better for performance.

 But it also depends on the compexity of your consumer widgets. If they are heavy to initialize you might want to consider using Riverpods StreamProvider, so you also dont have to pass so many references around.

**Reply by frankvollebregt** [reply]
Thank you! I've been relying on the `provider` package for simple state management until now, but I'll reconsider using riverpod.

I was indeed experiencing some poor performance, but from what I can tell from your comment, this may be due to the many nested async events that trigger rebuilds. I use the [multiple_stream_builder](https://pub.dev/packages/multiple_stream_builder) package, but internally I think it might be nesting multiple of these event listeners. As a result, some of the smaller "inner" widgets rely on around 4-6 different stream builders, each listening to up to 6 tables...

Are you aware of this having any impact on the query times of Drift? The reason I ask is because compared to a small test app with the same data model, my database operations take a long time (unacceptably long on slower/older devices). See also my [earlier discussion here](https://github.com/simolus3/drift/discussions/2335).

**Reply by FaFre** [reply]
Riverpod brings a lot of neat additions, its definitely worth the upgrade.

From my experience it's best to reduce the listened streams to a bare minimum, especially when you have a lot of events constantly or in bursts. You cannot really control the order when adding to multiple streams at around the same time, the async queue can be pretty unpredictable and can lead to a lot of unnecessary rebuilds. From my experience its almost always better to rebuild the complete tree which feels less performant but turned out faster in most of my cases. Also static or "less dynamic" widgets are much easier to maintain. I also make a lot of use of the rxdart package, it brings some really neat functionalities when working with streams. Check that out when you work a lot with streams.

Don't use isolates when transferring a lot of data (SELECT's!). The isolated memory will most likely kill all performance benefits by a multitude, because the copy and SendPort operations are really slow. In the vast majority of cases isolates will lead to a much worse performance, unless you do large updates or long running queries of at least a couple of seconds. The only real benefit of isolates in this scenario is non-blocking ui, but you pay with worse performance.

Your best friends for performance questions are the dart dev tools. Start the CPU profiler, run the operation a couple of times and investigate in the CPU flame chart. There you will see where the time is spend. That tool is gold!

**Reply by frankvollebregt** [reply]
Thank you so much for the insights and tips!
